### ISSUE TITLE
Preferences save Guru Meditation Error

### ISSUE THREAD

**Issue by petr-simek** [esphome/issues#495]
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

linux, pip, version 1.13.6
**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32S

**Affected component:**

Custom

**Description of problem:**
I created custom component for Motor, component used encoder for position and break. It work ok but if I used `preferences` I got error. Preferences is used for store last encoder position. It is able to store encoder state 2-5 times than it break.
Can you help me what I do wrong?

**Problem-relevant YAML-configuration entries:**
```yaml
sensor:
  - platform: rotary_encoder
    id: rotary_encoder_
    name: "Rotary Encoder"
    pin_a: 4
    pin_b: 16

```
no yaml
**Logs (if applicable):**

```
[07:15:33]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:33]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:33]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:33]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:34]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:34]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:34]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:34]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:34]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]Memory dump at 0x40162294: bad00bad bad00bad bad00bad
[07:15:34]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[07:15:34]MemoGuru Meditation Error: Core  1 panic'ed (Unhandled debug exception)
[07:15:34]Debug exception reason: Stack canary watchpoint triggered (loop�      @4) 
```
``` 
#include <utility>

#include "motor.h"


static unsigned int position_abs(unsigned int a, unsigned int b) {
    return a >= b ? a - b : b - a;
}

static const char *motor_Tag = "motor";

void Motor::off() {
    esphome::ESP_LOGD(motor_Tag, "off 1");
    l_enable->digital_write(false);
    r_enable->digital_write(false);
    esphome::ESP_LOGD(motor_Tag, "off 2");
    l_pwm->set_level(0);
    r_pwm->set_level(0);
    state = OFF;
    speed = 0;
    esphome::ESP_LOGD(motor_Tag, "off 3");
    float state = encoder->state;
    esphome::ESP_LOGD(motor_Tag, "off 4 %d", rtc.is_initialized());
//FAILED HERE  this->rtc.save(&state);
    this->rtc.save(&state);
    esphome::ESP_LOGD(motor_Tag, "off 5");
}

void Motor::setup() {
    Component::setup();
    object_id_ = "motor_sensor";
    encoder->state = getInitialEncoder();
    off();
}

void Motor::update() {
    if (state == OFF || disabled) {
        return;
    }
    encoder_break();
}

float Motor::getInitialEncoder() {
    rtc = esphome::global_preferences.make_preference<float>(this->get_object_id_hash());
    float initial_state;
    if (!rtc.load(&initial_state))
        return 0;
    return initial_state;
}

void Motor::encoder_break() {
    unsigned long now = millis();
    unsigned long diff = get_period(last_encoder_move, now);
    unsigned long encoder_state = encoder->state;
    if (diff >= 300 && position_abs(last_position, encoder_state) <= 0) {
        esphome::ESP_LOGD(motor_Tag, "off encoder_break");
        off();
        return;
    } else if (diff >= 300) {
        last_encoder_move = now;
        last_position = encoder_state;
    }
}

uint32_t Motor::hash_base() { return 1985890752UL; }
```

```
class Motor : public esphome::PollingComponent, esphome::Nameable {
public:
    Motor(uint32_t updateInterval, esphome::GPIOPin *rEnable, esphome::GPIOPin *lEnable,
          esphome::ledc::LEDCOutput *rPwm, esphome::ledc::LEDCOutput *lPwm,
          esphome::rotary_encoder::RotaryEncoderSensor *encoder);

    void off();

    void update() override;

    void setup() override;

private:
    float getInitialEncoder();

    void encoder_break();

    uint32_t hash_base() override ;

     esphome::rotary_encoder::RotaryEncoderSensor *encoder;
    esphome::ESPPreferenceObject rtc;
};
```
**Additional information and things you've tried:**

**Comment by OttoWinter** [comment]
Please post your entire code - including how you register and create the component. The error indicates that you're calling encoder_break from an interrupt, which will lead to an error.

**Comment by petr-simek** [comment]
Component is create and register in main.cpp 
```
 motor = new Motor(10, new GPIOPin(22, OUTPUT), new GPIOPin(15, OUTPUT), output_r_pwm, output_l_pwm, rotary_encoder_);
   App.register_component(motor);
```
I used button for run/stop motor like this
```
  - platform: gpio
    pin:
      number: 33
      mode: INPUT_PULLUP
      inverted: True
    name: "up"
    on_click:
      then:
        - lambda: |-
            motor->dir_ccw();
```
I use lambda for `motor->dir_ccw();` which has 2 functions. First push enable motor and second push stop motor and when motor is stop it call `this->rtc.save(&state);` ` esphome::ESPPreferenceObject rtc;`. 
Is this function run from interrupt?

```yml
esphome:
  name: stul
  platform: ESP32
  board: esp32dev
  includes:
    - esphomelib-component/Button.h
    - esphomelib-component/Button.cpp
    - esphomelib-component/JC_Button.h
    - esphomelib-component/JC_Button.cpp

wifi:
  ssid: 'Malarodinka_iot'
  password: 'lalamrdi1X'

# Enable logging
logger:
  level: DEBUG
api:
ota:

esp32_touch:
  setup_mode: False

binary_sensor:
  - platform: gpio
    pin:
      number: 33
      mode: INPUT_PULLUP
      inverted: True
    name: "up"
    on_click:
      then:
        - lambda: |-
            motor->dir_ccw();
  - platform: gpio
    pin:
      number: 5
      mode: INPUT_PULLUP
      inverted: True
    name: "down"
    on_click:
      then:
        - lambda: |-
            motor->dir_cw();
#  - platform: gpio
#    pin:
#      number: 25
#      mode: INPUT_PULLUP
#    name: "25-d"
#  - platform: gpio
#    pin:
#      number: 26
#      mode: INPUT_PULLUP
#    name: "26-d"
  - platform: gpio
    pin:
      number: 32
      mode: INPUT_PULLUP
      inverted: True
    name: "position_1"
    on_click:
      then:
        - lambda: |-
            motor->set_position(1);
  - platform: gpio
    pin:
      number: 27
      mode: INPUT_PULLUP
      inverted: True
    name: "position_2"
    on_click:
      then:
        - lambda: |-
            motor->set_position(2);
  - platform: gpio
    pin:
      number: 18
      mode: INPUT_PULLUP
      inverted: True
    name: "position_3"
    on_click:
      then:
        - lambda: |-
            motor->set_position(3);

dallas:
  - pin: 23

sensor:
  - platform: dallas
    index: 0
    name: "Stul teplomer"
  - platform: wifi_signal
    name: "Stul signal"
    update_interval: 60s
  - platform: rotary_encoder
    id: rotary_encoder_
    name: "Rotary Encoder"
    pin_a: 4
    pin_b: 16

custom_component:
  - lambda: |-
      auto my_sensor = new ButtonLight(14, id(svetlo_gauc));
      return {my_sensor};
switch:
  - platform: restart
    name: "Stul Restart"

text_sensor:
  - platform: wifi_info
    ip_address:
      name: 'Stul IP'

light:
  - platform: monochromatic
    name: "Svetlo Stul"
    output: output_component1
    id: svetlo_gauc

output:
  - platform: ledc
    id: output_component1
    pin: 19

  - platform: ledc
    id: output_r_pwm
    pin: 17
  - platform: ledc
    id: output_l_pwm
    pin: 2
```

Main.ccp
```
// Auto generated code by esphome
// ========== AUTO GENERATED INCLUDE BLOCK BEGIN ===========
#include "esphome.h"
using namespace esphome;
using namespace binary_sensor;
using namespace switch_;
using namespace text_sensor;
using namespace light;
logger::Logger *logger_logger;
wifi::WiFiComponent *wifi_wificomponent;
ota::OTAComponent *ota_otacomponent;
api::APIServer *api_apiserver;
using namespace sensor;
esp32_touch::ESP32TouchComponent *esp32_touch_esp32touchcomponent;
gpio::GPIOBinarySensor *gpio_gpiobinarysensor;
gpio::GPIOBinarySensor *gpio_gpiobinarysensor_2;
gpio::GPIOBinarySensor *gpio_gpiobinarysensor_3;
gpio::GPIOBinarySensor *gpio_gpiobinarysensor_4;
gpio::GPIOBinarySensor *gpio_gpiobinarysensor_5;
wifi_signal::WiFiSignalSensor *wifi_signal_wifisignalsensor;
rotary_encoder::RotaryEncoderSensor *rotary_encoder_;
restart::RestartSwitch *restart_restartswitch;
wifi_info::IPAddressWiFiInfo *wifi_info_ipaddresswifiinfo;
monochromatic::MonochromaticLightOutput *monochromatic_monochromaticlightoutput;
light::LightState *svetlo_gauc;
using namespace output;
binary_sensor::ClickTrigger *binary_sensor_clicktrigger;
Automation<> *automation;
binary_sensor::ClickTrigger *binary_sensor_clicktrigger_2;
Automation<> *automation_2;
binary_sensor::ClickTrigger *binary_sensor_clicktrigger_3;
Automation<> *automation_3;
binary_sensor::ClickTrigger *binary_sensor_clicktrigger_4;
Automation<> *automation_4;
binary_sensor::ClickTrigger *binary_sensor_clicktrigger_5;
Automation<> *automation_5;
dallas::ESPOneWire *dallas_esponewire;
dallas::DallasComponent *dallas_dallascomponent;
ledc::LEDCOutput *output_component1;
ledc::LEDCOutput *output_r_pwm;
ledc::LEDCOutput *output_l_pwm;
LambdaAction<> *lambdaaction;
LambdaAction<> *lambdaaction_2;
LambdaAction<> *lambdaaction_3;
LambdaAction<> *lambdaaction_4;
LambdaAction<> *lambdaaction_5;
dallas::DallasTemperatureSensor *dallas_dallastemperaturesensor;
#include "Button.h"
#include "JC_Button.h"
// ========== AUTO GENERATED INCLUDE BLOCK END ==========="
#include "motor.h"

Motor * motor;

void setup() {
  // ===== DO NOT EDIT ANYTHING BELOW THIS LINE =====
  // ========== AUTO GENERATED CODE BEGIN ===========
  // esphome:
  //   name: stul
  //   platform: ESP32
  //   board: esp32dev
  //   includes:
  //   - esphomelib-component/Button.h
  //   - esphomelib-component/Button.cpp
  //   - esphomelib-component/JC_Button.h
  //   - esphomelib-component/JC_Button.cpp
  //   build_path: stul
  //   platformio_options: {}
  //   arduino_version: espressif32@1.6.0
  //   libraries: []
  App.pre_setup("stul", __DATE__ ", " __TIME__);
  // binary_sensor:
  // switch:
  // text_sensor:
  // light:
  // logger:
  //   level: DEBUG
  //   logs: {}
  //   baud_rate: 115200
  //   tx_buffer_size: 512
  //   hardware_uart: UART0
  //   id: logger_logger
  logger_logger = new logger::Logger(115200, 512, logger::UART_SELECTION_UART0);
  logger_logger->pre_setup();
  App.register_component(logger_logger);
  // wifi:
  //   fast_connect: false
  //   domain: .local
  //   power_save_mode: NONE
  //   reboot_timeout: 5min
  //   id: wifi_wificomponent
  //   networks:
  //   - password: lalamrdi1X
  //     ssid: Malarodinka_iot
  //     id: wifi_wifiap
  //   use_address: stul.local
  wifi_wificomponent = new wifi::WiFiComponent();
  wifi_wificomponent->set_use_address("stul.local");
  wifi::WiFiAP wifi_wifiap = wifi::WiFiAP();
  wifi_wifiap.set_ssid("Malarodinka_iot");
  wifi_wifiap.set_password("lalamrdi1X");
  wifi_wificomponent->add_sta(wifi_wifiap);
  wifi_wificomponent->set_reboot_timeout(300000);
  wifi_wificomponent->set_power_save_mode(wifi::WIFI_POWER_SAVE_NONE);
  wifi_wificomponent->set_fast_connect(false);
  App.register_component(wifi_wificomponent);
  // ota:
  //   password: ''
  //   id: ota_otacomponent
  //   safe_mode: true
  //   port: 3232
  ota_otacomponent = new ota::OTAComponent();
  ota_otacomponent->set_port(3232);
  ota_otacomponent->set_auth_password("");
  ota_otacomponent->start_safe_mode();
  App.register_component(ota_otacomponent);
  // api:
  //   reboot_timeout: 5min
  //   password: ''
  //   id: api_apiserver
  //   port: 6053
  api_apiserver = new api::APIServer();
  App.register_component(api_apiserver);
  // sensor:
  api_apiserver->set_port(6053);
  api_apiserver->set_password("");
  api_apiserver->set_reboot_timeout(300000);
  // esp32_touch:
  //   setup_mode: false
  //   low_voltage_reference: 0.5V
  //   iir_filter: 0ms
  //   high_voltage_reference: 2.7V
  //   voltage_attenuation: 0V
  //   sleep_duration: 27306us
  //   id: esp32_touch_esp32touchcomponent
  //   measurement_duration: 8192us
  esp32_touch_esp32touchcomponent = new esp32_touch::ESP32TouchComponent();
  App.register_component(esp32_touch_esp32touchcomponent);
  // binary_sensor.gpio:
  //   platform: gpio
  //   pin:
  //     number: 33
  //     mode: INPUT_PULLUP
  //     inverted: true
  //   name: up
  //   on_click:
  //   - then:
  //     - lambda: !lambda |-
  //         motor->dir_ccw();
  //       type_id: lambdaaction
  //     min_length: 50ms
  //     automation_id: automation
  //     max_length: 350ms
  //     trigger_id: binary_sensor_clicktrigger
  //   id: gpio_gpiobinarysensor
  gpio_gpiobinarysensor = new gpio::GPIOBinarySensor();
  App.register_component(gpio_gpiobinarysensor);
  // binary_sensor.gpio:
  //   platform: gpio
  //   pin:
  //     number: 5
  //     mode: INPUT_PULLUP
  //     inverted: true
  //   name: down
  //   on_click:
  //   - then:
  //     - lambda: !lambda |-
  //         motor->dir_cw();
  //       type_id: lambdaaction_2
  //     min_length: 50ms
  //     automation_id: automation_2
  //     max_length: 350ms
  //     trigger_id: binary_sensor_clicktrigger_2
  //   id: gpio_gpiobinarysensor_2
  gpio_gpiobinarysensor_2 = new gpio::GPIOBinarySensor();
  App.register_component(gpio_gpiobinarysensor_2);
  // binary_sensor.gpio:
  //   platform: gpio
  //   pin:
  //     number: 32
  //     mode: INPUT_PULLUP
  //     inverted: true
  //   name: position_1
  //   on_click:
  //   - then:
  //     - lambda: !lambda |-
  //         motor->set_position(1);
  //       type_id: lambdaaction_3
  //     min_length: 50ms
  //     automation_id: automation_3
  //     max_length: 350ms
  //     trigger_id: binary_sensor_clicktrigger_3
  //   id: gpio_gpiobinarysensor_3
  gpio_gpiobinarysensor_3 = new gpio::GPIOBinarySensor();
  App.register_component(gpio_gpiobinarysensor_3);
  // binary_sensor.gpio:
  //   platform: gpio
  //   pin:
  //     number: 27
  //     mode: INPUT_PULLUP
  //     inverted: true
  //   name: position_2
  //   on_click:
  //   - then:
  //     - lambda: !lambda |-
  //         motor->set_position(2);
  //       type_id: lambdaaction_4
  //     min_length: 50ms
  //     automation_id: automation_4
  //     max_length: 350ms
  //     trigger_id: binary_sensor_clicktrigger_4
  //   id: gpio_gpiobinarysensor_4
  gpio_gpiobinarysensor_4 = new gpio::GPIOBinarySensor();
  App.register_component(gpio_gpiobinarysensor_4);
  // binary_sensor.gpio:
  //   platform: gpio
  //   pin:
  //     number: 18
  //     mode: INPUT_PULLUP
  //     inverted: true
  //   name: position_3
  //   on_click:
  //   - then:
  //     - lambda: !lambda |-
  //         motor->set_position(3);
  //       type_id: lambdaaction_5
  //     min_length: 50ms
  //     automation_id: automation_5
  //     max_length: 350ms
  //     trigger_id: binary_sensor_clicktrigger_5
  //   id: gpio_gpiobinarysensor_5
  gpio_gpiobinarysensor_5 = new gpio::GPIOBinarySensor();
  App.register_component(gpio_gpiobinarysensor_5);
  // dallas:
  //   pin:
  //     mode: INPUT
  //     number: 23
  //     inverted: false
  //   update_interval: 60s
  //   id: dallas_dallascomponent
  //   one_wire_id: dallas_esponewire
  // sensor.dallas:
  //   platform: dallas
  //   index: 0
  //   name: Stul teplomer
  //   unit_of_measurement: °C
  //   id: dallas_dallastemperaturesensor
  //   dallas_id: dallas_dallascomponent
  //   icon: mdi:thermometer
  //   resolution: 12
  //   accuracy_decimals: 1
  // sensor.wifi_signal:
  //   platform: wifi_signal
  //   name: Stul signal
  //   update_interval: 60s
  //   unit_of_measurement: dB
  //   id: wifi_signal_wifisignalsensor
  //   icon: mdi:wifi
  //   accuracy_decimals: 0
  wifi_signal_wifisignalsensor = new wifi_signal::WiFiSignalSensor();
  wifi_signal_wifisignalsensor->set_update_interval(60000);
  App.register_component(wifi_signal_wifisignalsensor);
  // sensor.rotary_encoder:
  //   platform: rotary_encoder
  //   id: rotary_encoder_
  //   name: Rotary Encoder
  //   pin_a:
  //     mode: INPUT
  //     number: 4
  //     inverted: false
  //   pin_b:
  //     mode: INPUT
  //     number: 16
  //     inverted: false
  //   unit_of_measurement: steps
  //   icon: mdi:rotate-right
  //   resolution: 1
  //   accuracy_decimals: 0
  rotary_encoder_ = new rotary_encoder::RotaryEncoderSensor();
  App.register_component(rotary_encoder_);
  // custom_component:
  //   lambda: !lambda |-
  //     auto my_sensor = new ButtonLight(14, id(svetlo_gauc));
  //     return {my_sensor};
  //   id: custom_component_customcomponentconstructor
  // switch.restart:
  //   platform: restart
  //   name: Stul Restart
  //   id: restart_restartswitch
  //   icon: mdi:restart
  restart_restartswitch = new restart::RestartSwitch();
  App.register_component(restart_restartswitch);
  // text_sensor.wifi_info:
  //   platform: wifi_info
  //   ip_address:
  //     name: Stul IP
  //     id: wifi_info_ipaddresswifiinfo
  wifi_info_ipaddresswifiinfo = new wifi_info::IPAddressWiFiInfo();
  App.register_component(wifi_info_ipaddresswifiinfo);
  // light.monochromatic:
  //   platform: monochromatic
  //   name: Svetlo Stul
  //   output: output_component1
  //   id: svetlo_gauc
  //   output_id: monochromatic_monochromaticlightoutput
  //   gamma_correct: 2.8
  //   restore_mode: RESTORE_DEFAULT_OFF
  //   default_transition_length: 1s
  monochromatic_monochromaticlightoutput = new monochromatic::MonochromaticLightOutput();
  svetlo_gauc = new light::LightState("Svetlo Stul", monochromatic_monochromaticlightoutput);
  App.register_light(svetlo_gauc);
  App.register_component(svetlo_gauc);
  // output:
  // output.ledc:
  //   platform: ledc
  //   id: output_component1
  //   pin:
  //     mode: OUTPUT
  //     number: 19
  //     inverted: false
  //   frequency: 1000.0
  //   bit_depth: 15
  // output.ledc:
  //   platform: ledc
  //   id: output_r_pwm
  //   pin:
  //     mode: OUTPUT
  //     number: 17
  //     inverted: false
  //   frequency: 1000.0
  //   bit_depth: 15
  // output.ledc:
  //   platform: ledc
  //   id: output_l_pwm
  //   pin:
  //     mode: OUTPUT
  //     number: 2
  //     inverted: false
  //   frequency: 1000.0
  //   bit_depth: 15
  esp32_touch_esp32touchcomponent->set_setup_mode(false);
  esp32_touch_esp32touchcomponent->set_iir_filter(0);
  esp32_touch_esp32touchcomponent->set_sleep_duration(4096);
  esp32_touch_esp32touchcomponent->set_measurement_duration(65535);
  esp32_touch_esp32touchcomponent->set_low_voltage_reference(TOUCH_LVOLT_0V5);
  esp32_touch_esp32touchcomponent->set_high_voltage_reference(TOUCH_HVOLT_2V7);
  esp32_touch_esp32touchcomponent->set_voltage_attenuation(TOUCH_HVOLT_ATTEN_0V);
  App.register_binary_sensor(gpio_gpiobinarysensor);
  gpio_gpiobinarysensor->set_name("up");
  binary_sensor_clicktrigger = new binary_sensor::ClickTrigger(gpio_gpiobinarysensor, 50, 350);
  automation = new Automation<>(binary_sensor_clicktrigger);
  App.register_binary_sensor(gpio_gpiobinarysensor_2);
  gpio_gpiobinarysensor_2->set_name("down");
  binary_sensor_clicktrigger_2 = new binary_sensor::ClickTrigger(gpio_gpiobinarysensor_2, 50, 350);
  automation_2 = new Automation<>(binary_sensor_clicktrigger_2);
  App.register_binary_sensor(gpio_gpiobinarysensor_3);
  gpio_gpiobinarysensor_3->set_name("position_1");
  binary_sensor_clicktrigger_3 = new binary_sensor::ClickTrigger(gpio_gpiobinarysensor_3, 50, 350);
  automation_3 = new Automation<>(binary_sensor_clicktrigger_3);
  App.register_binary_sensor(gpio_gpiobinarysensor_4);
  gpio_gpiobinarysensor_4->set_name("position_2");
  binary_sensor_clicktrigger_4 = new binary_sensor::ClickTrigger(gpio_gpiobinarysensor_4, 50, 350);
  automation_4 = new Automation<>(binary_sensor_clicktrigger_4);
  App.register_binary_sensor(gpio_gpiobinarysensor_5);
  gpio_gpiobinarysensor_5->set_name("position_3");
  binary_sensor_clicktrigger_5 = new binary_sensor::ClickTrigger(gpio_gpiobinarysensor_5, 50, 350);
  automation_5 = new Automation<>(binary_sensor_clicktrigger_5);
  dallas_esponewire = new dallas::ESPOneWire(new GPIOPin(23, INPUT, false));
  dallas_dallascomponent = new dallas::DallasComponent(dallas_esponewire);
  dallas_dallascomponent->set_update_interval(60000);
  App.register_component(dallas_dallascomponent);
  App.register_sensor(wifi_signal_wifisignalsensor);
  wifi_signal_wifisignalsensor->set_name("Stul signal");
  wifi_signal_wifisignalsensor->set_unit_of_measurement("dB");
  wifi_signal_wifisignalsensor->set_icon("mdi:wifi");
  wifi_signal_wifisignalsensor->set_accuracy_decimals(0);
  App.register_sensor(rotary_encoder_);
  rotary_encoder_->set_name("Rotary Encoder");
  rotary_encoder_->set_unit_of_measurement("steps");
  rotary_encoder_->set_icon("mdi:rotate-right");
  rotary_encoder_->set_accuracy_decimals(0);
  App.register_switch(restart_restartswitch);
  restart_restartswitch->set_name("Stul Restart");
  restart_restartswitch->set_icon("mdi:restart");
  App.register_text_sensor(wifi_info_ipaddresswifiinfo);
  wifi_info_ipaddresswifiinfo->set_name("Stul IP");
  svetlo_gauc->set_restore_mode(light::LIGHT_RESTORE_DEFAULT_OFF);
  svetlo_gauc->set_default_transition_length(1000);
  svetlo_gauc->set_gamma_correct(2.800000f);
  output_component1 = new ledc::LEDCOutput(new GPIOPin(19, OUTPUT, false));
  App.register_component(output_component1);
  output_r_pwm = new ledc::LEDCOutput(new GPIOPin(17, OUTPUT, false));
  App.register_component(output_r_pwm);
  output_l_pwm = new ledc::LEDCOutput(new GPIOPin(2, OUTPUT, false));
  App.register_component(output_l_pwm);
  lambdaaction = new LambdaAction<>([=]() -> void {
      motor->dir_ccw();
  });
  lambdaaction_2 = new LambdaAction<>([=]() -> void {
      motor->dir_cw();
  });
  lambdaaction_3 = new LambdaAction<>([=]() -> void {
      motor->set_position(1);
  });
  lambdaaction_4 = new LambdaAction<>([=]() -> void {
      motor->set_position(2);
  });
  lambdaaction_5 = new LambdaAction<>([=]() -> void {
      motor->set_position(3);
  });
  custom_component::CustomComponentConstructor custom_component_customcomponentconstructor = custom_component::CustomComponentConstructor([=]() -> std::vector<Component *> {
      auto my_sensor = new ButtonLight(14, svetlo_gauc);
      return {my_sensor};
  });
  svetlo_gauc->add_effects({});
  dallas_dallastemperaturesensor = dallas_dallascomponent->get_sensor_by_index(0, 12);
  App.register_sensor(dallas_dallastemperaturesensor);
  dallas_dallastemperaturesensor->set_name("Stul teplomer");
  dallas_dallastemperaturesensor->set_unit_of_measurement("\302\260C");
  dallas_dallastemperaturesensor->set_icon("mdi:thermometer");
  dallas_dallastemperaturesensor->set_accuracy_decimals(1);
  rotary_encoder_->set_pin_a(new GPIOPin(4, INPUT, false));
  output_component1->set_frequency(1000.000000f);
  output_component1->set_bit_depth(15);
  output_r_pwm->set_frequency(1000.000000f);
  output_r_pwm->set_bit_depth(15);
  output_l_pwm->set_frequency(1000.000000f);
  output_l_pwm->set_bit_depth(15);
  rotary_encoder_->set_pin_b(new GPIOPin(16, INPUT, false));
  rotary_encoder_->set_resolution(rotary_encoder::ROTARY_ENCODER_1_PULSE_PER_CYCLE);
  automation->add_actions({lambdaaction});
  automation_2->add_actions({lambdaaction_2});
  automation_3->add_actions({lambdaaction_3});
  automation_4->add_actions({lambdaaction_4});
  automation_5->add_actions({lambdaaction_5});
  monochromatic_monochromaticlightoutput->set_output(output_component1);
  gpio_gpiobinarysensor->set_pin(new GPIOPin(33, INPUT_PULLUP, true));
  gpio_gpiobinarysensor_2->set_pin(new GPIOPin(5, INPUT_PULLUP, true));
  gpio_gpiobinarysensor_3->set_pin(new GPIOPin(32, INPUT_PULLUP, true));
  gpio_gpiobinarysensor_4->set_pin(new GPIOPin(27, INPUT_PULLUP, true));
  gpio_gpiobinarysensor_5->set_pin(new GPIOPin(18, INPUT_PULLUP, true));
  // =========== AUTO GENERATED CODE END ============
  // ========= YOU CAN EDIT AFTER THIS LINE =========


   motor = new Motor(10, new GPIOPin(22, OUTPUT), new GPIOPin(15, OUTPUT), output_r_pwm, output_l_pwm, rotary_encoder_);
   App.register_component(motor);
  App.setup();
}

void loop() {
  App.loop();
}

```

motor.cpp
```
#include <utility>

#include "motor.h"


static unsigned int position_abs(unsigned int a, unsigned int b) {
    return a >= b ? a - b : b - a;
}

static const char *motor_Tag = "motor";

void Motor::off() {
    esphome::ESP_LOGD(motor_Tag, "off 1");
    l_enable->digital_write(false);
    r_enable->digital_write(false);
    esphome::ESP_LOGD(motor_Tag, "off 2");
    l_pwm->set_level(0);
    r_pwm->set_level(0);
    state = OFF;
    speed = 0;
    esphome::ESP_LOGD(motor_Tag, "off 3");
    float state = encoder->state;
    esphome::ESP_LOGD(motor_Tag, "off 4 %d", rtc.is_initialized());
    this->rtc.save(&state);
    esphome::ESP_LOGD(motor_Tag, "off 5");
}

void Motor::dir_cw() {
    if (state != OFF) {
        esphome::ESP_LOGD(motor_Tag, "off dir_cw");
        off();
        return;
    }
    last_tick = millis();
    last_encoder_move = millis();
    state = CW;
    next_position = -999999;
    esphome::ESP_LOGD(motor_Tag, "dir_cw");
}

void Motor::dir_ccw() {
    if (state != OFF) {
        esphome::ESP_LOGD(motor_Tag, "off dir_ccw");
        off();
        return;
    }
    last_tick = millis();
    last_encoder_move = millis();
    state = CCW;
    next_position = 999999;
    esphome::ESP_LOGD(motor_Tag, "dir_ccw");
}

void Motor::set_position(int i) {
    if (state != OFF) {
        esphome::ESP_LOGD(motor_Tag, "off set_position");
        off();
        return;
    }
    switch (i) {
        case 1:
            next_position = position_first;
            state = FIX;
            break;
        case 2:
            next_position = position_second;
            state = FIX;
            break;
        case 3:
            next_position = position_third;
            state = FIX;
            break;
        default:
            esphome::ESP_LOGD(motor_Tag, "set positio %d state off", i);
            state = OFF;
    }
    esphome::ESP_LOGD(motor_Tag, "set positio %d state FIX", i);
    last_tick = millis();
    last_encoder_move = millis();
}

void Motor::disable() {
    disabled = true;
    off();
}

void Motor::setSpeed(MotorState state, int speed) {
    if (state == OFF) {
        esphome::ESP_LOGD(motor_Tag, "setSpeed off");
        return;
    }
    this->speed = speed < L_MAX_SPEED ? speed : L_MAX_SPEED;
    esphome::ESP_LOGD(motor_Tag, "set speed %d", this->speed);

    r_enable->digital_write(true);
    l_enable->digital_write(true);

    if (state == CCW) {
        l_pwm->set_level(0);
        r_pwm->set_level(this->speed);
    } else if (state == CW) {
        l_pwm->set_level(this->speed);
        r_pwm->set_level(0);
    }
}

void Motor::setup() {
    Component::setup();
    r_enable->setup();
    l_enable->setup();
    r_pwm->setup();
    l_pwm->setup();
    disabled = false;
    object_id_ = "motor_sensor";
    encoder->state = getInitialEncoder();
    off();
}

void Motor::update() {
    if (state == OFF || disabled || (disabled_first && millis() < 10000)) {
        return;
    } else if (disabled_first) {
        disabled_first = false;
    }

    if (state == FIX) {
        if (next_position < encoder->state) {
            state = CW;
        } else {
            state = CCW;
        }
    }

    if ((state == CW && encoder->state <= next_position) ||
        (state == CCW && encoder->state >= next_position)) {
        esphome::ESP_LOGD(motor_Tag, "update condition off %d", speed);
        off();
        return;
    }

    encoder_break();

    unsigned long now = millis();
    unsigned long diff = get_period(last_tick, now);

    if (diff >= L_SPEED_STEP_DURATION) {
        setSpeed(state, speed + 1);
        last_tick = now;
    }
}

unsigned long get_period(const unsigned long last, const unsigned long next) {
    return next - last;
}

float Motor::getInitialEncoder() {
    rtc = esphome::global_preferences.make_preference<float>(this->get_object_id_hash());
    float initial_state;
    if (!rtc.load(&initial_state))
        return 0;
    return initial_state;
}

void Motor::encoder_break() {
    unsigned long now = millis();
    unsigned long diff = get_period(last_encoder_move, now);
    unsigned long encoder_state = encoder->state;
    if (diff >= 300 && position_abs(last_position, encoder_state) <= 0) {
        esphome::ESP_LOGD(motor_Tag, "off encoder_break");
        off();
        return;
    } else if (diff >= 300) {
        last_encoder_move = now;
        last_position = encoder_state;
    }
}

Motor::Motor(uint32_t updateInterval, esphome::GPIOPin *rEnable, esphome::GPIOPin *lEnable,
             esphome::ledc::LEDCOutput *rPwm, esphome::ledc::LEDCOutput *lPwm,
             esphome::rotary_encoder::RotaryEncoderSensor *encoder) : PollingComponent(updateInterval),
                                                                      r_enable(rEnable),
                                                                      l_enable(lEnable),
                                                                      r_pwm(rPwm),
                                                                      l_pwm(lPwm), encoder(encoder) {
}

uint32_t Motor::hash_base() { return 1985890752UL; }
```

motor.h
```
//
// Created by petr on 4.6.19.
//

#ifndef POKOJ_MOTOR_H
#define POKOJ_MOTOR_H

#include <esphome.h>
#include <limits.h>
#include "esphome/core/log.h"
typedef enum {
    CCW,
    CW,
    OFF,
    FIX
} MotorState;


unsigned long get_period(unsigned long last, unsigned long next);

#define MINIMUM_POS_CHANGE 8

class Motor : public esphome::PollingComponent, esphome::Nameable {
public:
    Motor(uint32_t updateInterval, esphome::GPIOPin *rEnable, esphome::GPIOPin *lEnable,
          esphome::ledc::LEDCOutput *rPwm, esphome::ledc::LEDCOutput *lPwm,
          esphome::rotary_encoder::RotaryEncoderSensor *encoder);

    void off();

    void dir_cw();

    void dir_ccw();

    void set_position(int i);

    void disable();;

    void update() override;

    void setup() override;

private:

    float getInitialEncoder();

    void encoder_break();

    uint32_t hash_base() override ;

    esphome::GPIOPin *r_enable;
    esphome::GPIOPin *l_enable;
    esphome::ledc::LEDCOutput *r_pwm;

    esphome::ledc::LEDCOutput *l_pwm;
    int L_MAX_SPEED = 1024;

    int L_SPEED_STEP_DURATION = 200;
    int position_first = 50;
    int position_second = 100;

    int position_third = 10;
    int speed = 0;

    esphome::rotary_encoder::RotaryEncoderSensor *encoder;
    MotorState state = OFF;
    bool disabled = false;
    bool disabled_first = true;
    long next_position = -1;
    long last_position = 0;
    void setSpeed(MotorState state, int speed);
    unsigned long last_tick = 0;

    unsigned long last_encoder_move = 0;

    esphome::ESPPreferenceObject rtc;
};


#endif //POKOJ_MOTOR_H

```

**Comment by OttoWinter** [comment]
Well I don't immediately see what would cause the issue you see with the crash, but I see a couple of other bugs:

 - Motor::setup() calls setup functions of the ledc objects - but setup is already called from them from their own component lifecycle - remove those calls.
 - Motor::setup also sets encoder->state - but encoder->state is only an internal value. Changing it won't affect the states published in the future. Haven't looked at the code too closely, but this might make the component save the state many hundred times per second, in which case I think the esp-idf nvs SDK may also throw issues.

Please also try using the `globals` component for this saved data instead of using your own rtc variable.

**Comment by petr-simek** [comment]
duplicit setups are removed. 
`encoder->state = custom_vule` is call only once in `Motor::setup()`. Is there any solution how to set init state?

How you mean this? `Please also try using the globals component for this saved data instead of using your own rtc variable.` 
I used same code as you used in `switch` and I don't know there is some other component which can be used for data store. Can you please navigate me where shoud I look?

**Comment by glmnet** [comment]
Search globals in docs. Is under automation

**Comment by petr-simek** [comment]
I rewrite preference(rtc) to globals and now it save value but then it throw same error as before.
```
[16:48:47]Guru Meditation Error: Core  1 panic'ed (Cache disabled but cached memory region accessed)
[16:48:47]Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
[16:48:47]Memory dump at 0x40161040: bad00bad bad00bad bad00bad
```
Interaction with global is only in one method
```
void Motor::_off() {
     l_enable->digital_write(false);
    r_enable->digital_write(false);
    ESP_LOGD(motor_Tag, "off 2");
    l_pwm->set_level(0);
    r_pwm->set_level(0);
    state = OFF;
    speed = 0;
    globalEncoderState->value() = encoder->state;
    ESP_LOGD(motor_Tag, "off 3");
}
```
Method is call only from Component#update() which is not call from an interrupt and only once in while. 

In my case even if I stop motor, rotary encoder do other few steps. When I postponed encoder state save (no more steps) it start working properly. Is it by accident or rotary encoder has some bug or something else?

```
void Motor::update() {
    if (!encoder_store) {
        save_encoder();
    }
    if (state == SHUTDOWN) {
        _off();
        return;
    }
.......
}
void Motor::_off() {
    l_enable->digital_write(false);
    r_enable->digital_write(false);
    l_pwm->set_level(0);
    r_pwm->set_level(0);
    state = OFF;
    speed = 0;
    encoder_store = false;
}

void Motor::save_encoder() {
    unsigned long now = millis();
    unsigned long diff = get_period(last_encoder_move, now);
    unsigned long encoder_state = encoder->state;
    if (diff >= 1000 && position_abs(last_position, encoder_state) == 0) {
        globalEncoderState->value() = encoder->state;
        encoder_store = true;
    }
}
```

**Comment by OttoWinter** [comment]
> I rewrite preference(rtc) to globals and now it save value but then it throw same error as before.

Can you try removing the globals saving and see if the error still occurs? I highly doubt it's the globals, I think I would have heard bug reports about it (including from myself) if there were a bug there.

> Method is call only from Component#update() which is not call from an interrupt and only once in while.

No, `update` is only called from non-ISR context.

> In my case even if I stop motor, rotary encoder do other few steps. When I postponed encoder state save (no more steps) it start working properly. Is it by accident or rotary encoder has some bug or something else?

I don't understand what you mean there. Could you try rephrasing and explaining with code?

**Comment by petr-simek** [comment]
> > I rewrite preference(rtc) to globals and now it save value but then it throw same error as before.
> 
> Can you try removing the globals saving and see if the error still occurs? I highly doubt it's the globals, I think I would have heard bug reports about it (including from myself) if there were a bug there.
> 
If I remove `globalEncoderState->value() = encoder->state;` it is work without error.

> > In my case even if I stop motor, rotary encoder do other few steps. When I postponed encoder state save (no more steps) it start working properly. Is it by accident or rotary encoder has some bug or something else?
> 
> I don't understand what you mean there. Could you try rephrasing and explaining with code?

1) press button
2) call _off()
3) Disconnect electricity from motor (It take few ms before motor actually stop. I don't know how to explain it, If you look at https://youtu.be/NFvMC3l3fGY?t=79 1:20-1:30)
4) Rotary encoder count other 8 steps
5) If encoder has same count for 1s call  `globalEncoderState->value() = encoder->state;`
6) ok

but if I changed it
1) press button
2) call _off()
3) Disconnect electricity from motor and call  `globalEncoderState->value() = encoder->state;`
4a) 1 of 3 attempts it count
4b) 2 of 3 attempts it failed with error

If you don't understand what I mean, I'm not surprised, my English is not very good. You helped me with my problem and we can close the issue.
